app/models/reducer.rb

```diff
@@ -40,22 +45,68 @@ def self.of_type(type)
 
   NoData = Class.new
 
-  def process(extracts)
+  def process(extracts, reductions=nil)
     light = Stoplight("reducer-#{id}") do
       grouped_extracts = ExtractGrouping.new(extracts, grouping).to_h
 
-      grouped_extracts.map do |key, grouped|
-        [key, reduction_data_for(extract_filter.filter(grouped))]
-      end.to_h
+      keys = { workflow_id: workflow_id, reducer_key: key }
+      keys[:subject_id] = extracts.first&.subject_id if reduce_by_subject?
+      keys[:user_id] = extracts.first&.user_id if reduce_by_user?
+
+      factory = if reduce_by_subject? then SubjectReduction elsif reduce_by_user? then UserReduction else nil end
+
+      new_reductions = grouped_extracts.map do |group_key, grouped|
+        keys[:subgroup] = group_key
+
+        reduction = if reductions.present?
+          reductions.where(keys).first_or_initialize
+        else
+          factory.new(keys)
+        end
+
+        filtered_extracts = extract_filter.filter(grouped)
```

Comment:
Extract filtering by counts is probably not going to work correctly, in hindsight. For example, if I have "first 5" set up, and the reducer has already seen 3 extracts, and 3 more extracts come in, it's not going to be able to know that it should drop one of them. I'm not sure if this is a problem that should be fixed with implementation or by not allowing that configuration.

Reply:
Adding, that implementing `0-n` range filtering is imaginable but difficult, but implementing `m-n` range filtering is a lot harder and needs some thought